packages/wrangler/src/output.ts

```diff
@@ -0,0 +1,95 @@
+import { randomBytes } from "node:crypto";
+import { appendFileSync } from "node:fs";
+import { resolve } from "node:path";
+import {
+	getOutputFileDirectoryFromEnv,
+	getOutputFilePathFromEnv,
+} from "./environment-variables/misc-variables";
+import { ensureDirectoryExistsSync } from "./utils/filesystem";
+
+/**
+ * Write an entry to the output file.
+ *
+ * Control whether (and where) to write this file by setting either
+ * `WRANGLER_OUTPUT_FILE_DIRECTORY` or `WRANGLER_OUTPUT_FILE_PATH`.
+ */
+export function writeOutput(entry: OutputEntry) {
+	if (outputFilePath === undefined) {
+		outputFilePath = getOutputFilePath();
+	}
+	if (outputFilePath !== null) {
+		ensureDirectoryExistsSync(outputFilePath);
+		const entryJSON = JSON.stringify({
+			...entry,
+			timestamp: new Date().toISOString(),
+		});
+		appendFileSync(outputFilePath, entryJSON + "\n");
+	}
+}
+
+// Only used internally for cleaning up tests
+export function clearOutputFilePath() {
+	outputFilePath = undefined;
+}
+
+let outputFilePath: string | null | undefined = undefined;
+function getOutputFilePath() {
+	const outputFilePathFromEnv = getOutputFilePathFromEnv();
+	if (outputFilePathFromEnv) {
+		return outputFilePathFromEnv;
+	}
+
+	const outputFileDirectoryFromEnv = getOutputFileDirectoryFromEnv();
+	if (outputFileDirectoryFromEnv) {
+		const date = new Date()
+			.toISOString()
+			.replaceAll(":", "-")
+			.replace(".", "_")
+			.replace("T", "_")
+			.replace("Z", "");
+
+		return resolve(
+			outputFileDirectoryFromEnv,
+			`wrangler-output-${date}-${randomBytes(3).toString("hex")}.json`
+		);
+	}
+
+	return null;
+}
+
+interface OutputEntryBase<T extends string> {
+	version: number;
```

Comment:
Might want to constrain the type here to `version: 1` to prevent any accidental mis-versioning

Reply:
I was just thinking that to myself!

Reply:
a981459e670155e2d623f997f44f5264cc81c76f

Reply:
I set it at the individual entry level rather than in the base, since they may want to evolve independently.